test/test_pid_reuse.c

```diff
@@ -0,0 +1,166 @@
+/* Test the reuse of PID by opening an NX window and then forking other process
+ * to see if one of the child can use the same PID of the first one with the
+ * window still open.
+ *
+ * Copyright (C) 2020 IBM Corporation
+ *
+ * Licenses for GPLv2 and Apache v2.0:
+ *
+ * GPLv2:
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ *
+ * Apache v2.0:
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/syscall.h>
+#include <assert.h>
+#include <errno.h>
+#include <pthread.h>
+#include "zlib.h"
+
+/* Globals shared by all threads. */
+z_stream c_stream;
+
+void* comp_file_multith (void *iterations)
+{
+	int err;
+	long *sleep_sec;
+
+	pid_t my_tid = syscall(SYS_gettid);
+	fprintf(stderr, "thread started: pid: %d\n", (int) my_tid);
+
+	/* Open the window. */
+	err = deflateInit(&c_stream, Z_DEFAULT_COMPRESSION);
+	if (err != 0) {
+		printf("nx_deflateInit err %d\n", err);
+		return (void *) 1;
+	}
+
+	sleep_sec = (long *) iterations;
+	sleep(sleep_sec[0]);
+
+	fprintf(stderr, "thread %d exits\n", (int) my_tid);
+	return (void *) 0;
+}
+
+void fork_children (int parent_pid) {
+	int child_id, i = 0;
+	int seconds = 10;
+	int children = 2048;
+
+	for (i=0; i < children; i++) {
+		child_id = fork();
```

Comment:
Is this explicitly using a double-fork trick to avoid zombies?

Reply:
No, it is forking 2048 times to see if any child uses the same pid as the parent.